src/common/webrtc/io.rs

```diff
@@ -9,88 +9,109 @@ use std::{
 };
 
 use anyhow::bail;
+use async_channel::{RecvError, SendError};
 use bytes::{Buf, BufMut, Bytes, BytesMut};
-use futures_lite::{AsyncRead, AsyncWrite, Future, FutureExt};
+use futures_lite::{future::Boxed, ready, AsyncRead, AsyncWrite, FutureExt};
 use smol::net::UdpSocket;
 
 #[derive(Debug)]
 pub struct IoPkt {
     addr: SocketAddrV4,
     payload: Bytes,
 }
-#[derive(Clone)]
+unsafe impl Send for IoPktChannel {}
+unsafe impl Sync for IoPktChannel {}
+
 pub struct IoPktChannel {
     rx: smol::channel::Receiver<IoPkt>,
     transport_tx: smol::channel::Sender<IoPkt>,
     tx: smol::channel::Sender<IoPkt>,
     ip: Option<SocketAddrV4>,
+    recv_operation: Option<Boxed<Result<IoPkt, RecvError>>>,
+    send_operation: Option<Boxed<Result<(), SendError<IoPkt>>>>,
     waker: Option<Waker>,
 }
 
+impl Clone for IoPktChannel {
+    fn clone(&self) -> Self {
+        Self {
+            rx: self.rx.clone(),
+            transport_tx: self.transport_tx.clone(),
+            tx: self.tx.clone(),
+            ip: self.ip,
+            recv_operation: None,
+            send_operation: None,
+            waker: None,
+        }
+    }
+}
+
 impl AsyncRead for IoPktChannel {
     fn poll_read(
         mut self: Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
         mut buf: &mut [u8],
     ) -> Poll<std::io::Result<usize>> {
-        let mut recv = self.rx.recv();
-        let fut = Box::pin(&mut recv).poll(cx);
-        match fut {
-            Poll::Ready(ret) => match ret {
-                Ok(pkt) => {
-                    let _ = self.ip.insert(pkt.addr);
-                    if buf.len() < pkt.payload.len() {
-                        // TODO(npm) remove the panic
-                        panic!(
-                            "expected buf len {} to be more than payload {}",
-                            buf.len(),
-                            pkt.payload.len()
-                        );
-                    }
-                    let len = pkt.payload.len();
-                    buf.put(pkt.payload);
-                    Poll::Ready(Ok(len))
+        if self.recv_operation.is_none() {
```

Comment:
Not sure I understand what you mean, we need to mutate the struct because each time a recv_operation complete (Poll::Ready) we need to create a new one for the next operation since we cannot re-poll completed futures

Reply:
wait, so where is recv_operation cleared out in the struct then?

Reply:
when we do take `let _ = self.recv_operation.take();` before returning the result